File: 19-remove-nth-node-from-end-of-list/19.go
package _9_remove_nth_node_from_end_of_list

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil {
		return nil
	}

	d := &ListNode{Next: head}

	second := d
	first := d
	for i := 0; i <= n; i++ {
		first = first.Next
	}

	for first != nil {
		second = second.Next
		first = first.Next
	}

	second.Next = second.Next.Next

	return d.Next
}

func PrintList(head *ListNode) {
	p := head
	max := 100
	c := 0
	for p != nil {
		fmt.Print(p.Val, "\t")
		p = p.Next
		c++
		if c > max {
			break
		}
	}
	fmt.Println()
}

File: c/src/main/19-remove-nth-node-from-end-of-list/19.c
#include "19.h"
#include <stdlib.h>

struct ListNode *removeNthFromEnd(struct ListNode *head, int n) {
    struct ListNode *p1 = head;
    struct ListNode *p2 = head;

    while (p1 != NULL) {
        if (n < 0) {
            p2 = p2->next;
        }
        n--;
        p1 = p1->next;
    }

    if (n == 0) {
        return head->next;
    }

    struct ListNode *tmp = p2->next;
    p2->next = p2->next->next;
    free(tmp);

    return head;
}